Infer 1.1.0

andrewvaughanj opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This is a (draft) PR to support Infer version 1.1.0. I have tested the `--build-from-source` on macOS Big Sur; I'm hoping that opening-up this PR might shake out any issues.

**Infer 1.1.0 is not yet released, and this is building a formula for the draft GitHub release of Infer**

Questions:

- [x] How do I generate and obtain the SHA for a bottle for Infer on Big Sur?
- [x] I guess I also need to update the SHAs for bottles for other versions of macOS -- how do I do this?

Check the [docs](https://github.com/Homebrew/brew/blob/master/docs/Bottles.md) maybe this is all automated and there's nothing for me to do?

andrewvaughanj commented:
This supersedes:

- https://github.com/Homebrew/homebrew-core/pull/62862
- https://github.com/Homebrew/homebrew-core/pull/62576

andrewvaughanj commented:
@jvillard do you think we need to add a dependency on Java for Infer? Right now, the tests are failing:

```
==> /usr/local/Cellar/infer/1.1.0/bin/infer --fail-on-issue -P -- javac FailingTest.java
Capturing in javac mode...
Usage Error:
*** Failed to execute compilation command: exited with code 1
*** Command: javac '@/private/tmp/infer-test-20210318-53605-1w2vpu7/infer-out/tmp/javac_args.tmp.deb4f4'
*** Output:
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.


*** Infer needs a working compilation command to run.


Error: infer: failed
An exception occurred within a child process:
  Minitest::Assertion: Expected: 2
  Actual: 1
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.14.4/lib/minitest/assertions.rb:183:in `assert'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.14.4/lib/minitest/assertions.rb:218:in `assert_equal'
/usr/local/Homebrew/Library/Homebrew/formula_assertions.rb:65:in `shell_output'
/Users/avj/clones/homebrew-core/Formula/infer.rb:132:in `block in <class:Infer>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1915:in `block (3 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1914:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:900:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1913:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2161:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1907:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:43:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:42:in `<main>'
```

Given we know we *definitely* have `clang`, maybe we should use a C-based test rather than a Java-based one? If not, I guess we should add a dependancy on Java for Infer in Homebrew.

carlocab commented:
It does declare a dependency on Java here:

https://github.com/Homebrew/homebrew-core/blob/df8f951542ee7d3e2aef28e29551214275bc9dad/Formula/infer.rb#L32

The problem seems to be that it can't find Java.

>this is all automated and there's nothing for me to do

This is correct.

andrewvaughanj commented:
@carlocab should be fixed in 9f655f0 -- could you tell me if this is the right way to achieve what we wanted?

SMillerDev commented on Formula/infer.rb:
```diff
@@ -25,14 +24,12 @@ class Infer < Formula
-  depends_on "ocaml-num" => :build
+  depends_on "ninja" => :build
   depends_on "opam" => :build
   depends_on "openjdk@8" => [:build, :test]
```
Does it not work with a newer version?

andrewvaughanj replied:
Good question -- @jvillard might know.

I am happy to try a local build with a more recent JDK and see what fails -- do you have a suggestion for a version that is more appropriate?

SMillerDev replied:
In homebrew we always try for the latest version of software.

andrewvaughanj replied:
Do you mean just dropping the `@8`?

jvillard replied:
Infer supports Java11 too, we could try bumping the version of the openjdk dependency.

andrewvaughanj replied:
Bumped it to `@11` -- testing locally.

andrewvaughanj replied:
JDK11 worked locally; changed is pushed.

SMillerDev commented on Formula/infer.rb:
```diff
@@ -70,39 +60,16 @@ def install
-
-    system "opam", "exec", "--", "./build-infer.sh", "all", "--yes", "--user-opam-switch"
-    system "opam", "exec", "--", "make", "install-with-libs"
+    system "./build-infer.sh", "all", "--yes"
+    system "make", "install-with-libs"
```
Does this build a custom clang as the file is suggesting? Because that would conflict with https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

andrewvaughanj replied:
No -- Infer *uses* `clang` internally to parse and reason about source code (it is a static analysis tool). The version of Infer is intrinsically tied to a particular revision of `clang`.

This Formula does not build `clang` and then use `clang` to build Infer -- it uses the system compiler to build Infer, and Infer's build process builds a version of `clang` that it then links against.

Does that make sense? Is this acceptable?

SMillerDev replied:
So, on macOS it would use `LLVM` to build `infer`, and in doing so it would build a `clang` that it infer would then use instead of the system tools?

@Homebrew/maintainers any opinions on this? I know we removed facebook projects for doing these kind of things before.

andrewvaughanj replied:
> in doing so it would build a clang that it infer would then use instead of the system tools?
>

Not really ... it builds a `libclang.dylib` that it then links against and uses for obtaining the AST of C/C++ code.

The built `clang` executable is never expected be used by an end user.

andrewvaughanj replied:
Just to be clearer:

> The version of Infer is intrinsically tied to a particular revision of clang.
>

It is not expected that Infer would work with the system version of LLVM. It fact, it is more expected that it *won't* work.

Infer has a set of bindings between LLVM and OCaml -- these bindings are hand-written and hand-maintained to meet a specific version of LLVM.

It is not possible to just swap-in another LLVM version and expect these bindings to be compatible.

MikeMcQuaid replied:
> The built `clang` executable is never expected be used by an end user.

In that case: let's not install it, please.

> It is not expected that Infer would work with the system version of LLVM. It fact, it is more expected that it _won't_ work.

Whether it's expected or not: can you confirm whether it does/does not work with:
- the system Clang
- the Clang from any of the `llvm` or `llvm@*` formulae

If it's tested and known to be definitely broken for all of those: ok. If not: let's try to use one of them.

SMillerDev replied:
> Not really ... it builds a libclang.dylib that it then links against and uses for obtaining the AST of C/C++ code.
> The built clang executable is never expected be used by an end user.

Doesn't this also mean that if there is no way to "avoid shipping outdated/insecure versions of software that is already a formula"?

andrewvaughanj replied:
> In that case: let's not install it, please.

I'm currently building the `@11` JDK build, so don't have an install to check this ... but are you saying just remove the whole `bin` folder from the install? We need the libs, but not the LLVM/`clang` executables.

> If it's tested and known to be definitely broken for all of those: ok. If not: let's try to use one of them.

I 100% agree with the suggestion/intent/etc. with this; I'm just now out of my depth to resolve this, as I don't know how to "trick" Infer to use an out-of-tree build of Clang.

andrewvaughanj replied:
> That makes it fundamentally conflict with the goals of a package manager.

I am not sure that's true -- some packages need dependencies at different/older versions; some packages need *patched* dependencies at different/versions.

If Homebrew doesn't allow this (and that's totally fine if you don't!), then it is very possible that Infer is fundamentally incompatible with Homebrew as it does need a *very specific* build of LLVM (and there's a chance it is also patched as well: https://github.com/facebook/infer/tree/master/facebook-clang-plugins/clang/src).

MikeMcQuaid replied:
> I'm currently building the `@11` JDK build, so don't have an install to check this ... but are you saying just remove the whole `bin` folder from the `clang` install? We need the libs, but not the LLVM/`clang` executables.

If we don't need them: yep. That would reduce my concerns considerably.

> If Homebrew doesn't allow this (and that's totally fine if you don't!), then it is very possible that Infer is fundamentally incompatible with Homebrew as it does need a _very specific_ build of LLVM (and there's a chance it is also patched as well: https://github.com/facebook/infer/tree/master/facebook-clang-plugins/clang/src).

We have found that multiple Facebook projects in the past are designed in such a way to make them very hard to package securely and reproducibly (and the upstream projects have been less than receptive in helping us).

andrewvaughanj replied:
> If we don't need them: yep. That would reduce my concerns considerably.

Should be fine; I'll try this once my `@11` JDK build has finished (which takes about ~2 hours per build 🤦)

carlocab replied:
Thanks very much for your work on this, @andrewvaughanj. When you get around to it, this line can be removed:

https://github.com/Homebrew/homebrew-core/blob/ed18465cacb1eb1d46df543ae7f0c59a1cdae71b/Formula/infer.rb#L22

(I do agree with the comments above re the superfluous `clang` executable and trying to see if we can link with an existing `libclang`, etc, though)

andrewvaughanj replied:
> When you get around to it, this line can be removed:

Already gone in my local, unpushed version 👼

jvillard replied:
Thanks for this interesting discussion!

I'm afraid deleting the clang binaries won't work: Infer needs them as the C/ObjC/C++ frontends are all based on running clang+the infer plugin to extract the AST of source files. This is how clang plugins in general work: they need clang to run together with the plugin. The version of clang used by infer in its current version is stock clang 11 + a handful of patches. Without these patches the Infer clang plugin doesn't work either as clang normally tries to do much more than just parsing even when we only need the AST. For this reason we cannot make infer build against the packaged llvm either, unfortunately.

I'd like to reiterate that the modified clang that Infer uses under the hood is well-hidden from users of Infer and isolated from the rest of system so it doesn't affect anything else. We regularly update the version of clang too so I don't think there are security concerns (I'm not sure I understand the reproducibility concerns, but perhaps that wasn't directed at this PR specifically).

MikeMcQuaid replied:
> The version of clang used by infer in its current version is stock clang 11 + a handful of patches.

Are there any efforts to get these patches upstreamed?



> I'd like to reiterate that the modified clang that Infer uses under the hood is well-hidden from users of Infer and isolated from the rest of system so it doesn't affect anything else.

Can you elaborate on how it's well-hidden/isolated?

andrewvaughanj replied:
> Can you elaborate on how it's well-hidden/isolated?

The live in a directory such as `/usr/local/Cellar/infer/1.1.0/lib/infer/facebook-clang-plugins/clang/install/bin` (which isn't placed on `$PATH`) -- they *are not* put into `/usr/local/bin`.

jvillard replied:
> Are there any efforts to get these patches upstreamed?

That's a good question. In their current form some of these patches are not mergeable as-is as they disable some clang functionalities(!) but I imagine we could find a common ground with clang's needs, perhaps via a command-line option. We are not clang devs but we'll try and find help to hopefully submit something upstream. Thanks for the suggestion!

carlocab replied:
Please do find a way to get these patches upstreamed in some form. Homebrew doesn't like third-party patches (they tend to break things in unexpected and sometimes very damaging ways), and it would be good for users to not have to have several `clang` installations on their system just to run the software they use.

andrewvaughanj commented:
How are we feeling about this PR now? There's an action point on #TeamInfer to look at getting the LLVM changes upstreamed, but does that block this PR?

If we're not blocked on the LLVM side of things, is there anything else pending that needs to be done to get this particular PR "over the line"?

carlocab commented on Formula/infer.rb:
```diff
@@ -70,39 +61,16 @@ def install
   end
 
   test do
+    java_home = Formula["openjdk@11"].opt_prefix
```
```suggestion
    ENV["JAVA_HOME"] = Formula["openjdk@11"].opt_prefix
```
Does this work? i.e. do we need to specify the full path to `javac` below with this in place?

andrewvaughanj replied:
Done!

carlocab commented on Formula/infer.rb:
```diff
@@ -164,7 +132,7 @@ class PassingTest {
 
-    shell_output("#{bin}/infer --fail-on-issue -P -- javac FailingTest.java", 2)
-    shell_output("#{bin}/infer --fail-on-issue -P -- javac PassingTest.java")
+    shell_output("#{bin}/infer --fail-on-issue -P -- #{java_home}/bin/javac FailingTest.java", 2)
+    shell_output("#{bin}/infer --fail-on-issue -P -- #{java_home}/bin/javac PassingTest.java")
```
Do these produce any output we can check with `assert_equal` or `assert_match`? Same with the C tests above.

andrewvaughanj replied:
So, I did this, but I don't know what the correct style is for Homebrew -- I just copied the whole output.

What's the "norm" for this kind of thing?

carlocab commented on Formula/infer.rb:
```diff
@@ -127,8 +95,30 @@ def install
 
-    shell_output("#{bin}/infer --fail-on-issue -P -- clang -c FailingTest.c", 2)
-    shell_output("#{bin}/infer --fail-on-issue -P -- clang -c PassingTest.c")
+    no_issues_output =
+      ["", "  No issues found  ", ""].join("\n")
```
```suggestion
    no_issues_output = "\n  No issues found  \n"
```
Note that `assert_match` tests whether there is a matching substring, so this should probably work too:
```ruby
no_issues_output = "No issues found"
```
However, `assert_equal` is better than `assert_match` wherever possible.

carlocab commented on Formula/infer.rb:
```diff
@@ -127,8 +95,30 @@ def install
+    no_issues_output =
+      ["", "  No issues found  ", ""].join("\n")
+
+    failing_c_output =
+      ["", "FailingTest.c:5: error: Null Dereference",
+       "  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.",
+       "  3. int main() {",
+       "  4.   int *s = NULL;",
+       "  5.   *s = 42;",
+       "       ^",
+       "  6. ",
+       "  7.   return 0;",
+       "",
+       "",
+       "Found 1 issue",
+       "          Issue Type(ISSUED_TYPE_ID): #",
+       "  Null Dereference(NULL_DEREFERENCE): 1",
+       ""].join("\n")
```
```suggestion
    failing_c_output = <<~EOS

      FailingTest.c:5: error: Null Dereference
        pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
        3. int main() {
        4.   int *s = NULL;
        5.   *s = 42;
             ^
        6.
        7.   return 0;
      
      
      Found 1 issue
                Issue Type(ISSUED_TYPE_ID): #
        Null Dereference(NULL_DEREFERENCE): 1

    EOS
```
Formatting might need work.

andrewvaughanj replied:
Okay, so I did this (manually) and now the tests pass, but `brew style` complains because of the formatting.

Any suggestions here?

carlocab replied:
Indent the heredoc as suggested, and drop the whitespace inside `main` in `FailingTest.c` so that we don't need `6. ` with the trailing whitespace.

carlocab commented on Formula/infer.rb:
```diff
@@ -164,7 +154,25 @@ class PassingTest {
 
-    shell_output("#{bin}/infer --fail-on-issue -P -- javac FailingTest.java", 2)
-    shell_output("#{bin}/infer --fail-on-issue -P -- javac PassingTest.java")
+    failing_java_output =
+      ["", "FailingTest.java:12: error: Null Dereference",
+       "  object `s` last assigned on line 11 could be null and is dereferenced at line 12.",
+       "  10.     int mayCauseNPE() {",
+       "  11.       String s = mayReturnNull(0);",
+       "  12. >     return s.length();",
+       "  13.     }",
+       "  14.   }",
+       "",
+       "",
+       "Found 1 issue",
+       "          Issue Type(ISSUED_TYPE_ID): #",
+       "  Null Dereference(NULL_DEREFERENCE): 1",
+       ""].join("\n")
```
Same comment as above.

MikeMcQuaid reviewed (commented):
I'm fine with this once @carlocab's suggested changes are made.

Bo98 commented:
Tagging this as "do not merge" because of:

> **Infer 1.1.0 is not yet released, and this is building a formula for the draft GitHub release of Infer**

Will untag once it's released.

Bo98 commented on Formula/infer.rb:
```diff
@@ -127,8 +95,33 @@ def install
-    shell_output("#{bin}/infer --fail-on-issue -P -- clang -c PassingTest.c")
+    no_issues_output = <<~EOS
+
+\ \ No issues found  
```
Thoughts on a "match" rather than a strict "equals" comparison so that we don't need to worry about that?

carlocab replied:
Doesn't
```ruby
no_issues_output = "\n  No issues found  \n"
```
work?

andrewvaughanj replied:
Oh, for that message, probably 😂 -- I was just going for consistency for the bigger ones, but maybe that wasn't a wise choice.

andrewvaughanj replied:
Done!

MikeMcQuaid replied:
Use `<<-EOS` without indentation if you need to preserve trailing spaces.

carlocab commented on Formula/infer.rb:
```diff
@@ -127,8 +95,30 @@ def install
+  4.   int *s = NULL;
+  5.   *s = 42;
+       ^
+  6. 
```
`brew audit` is probably going to complain about trailing whitespace here. You could just get rid of the whitespace in `FailingTest.c` to avoid the need for this line.

andrewvaughanj replied:
Ah! Good idea! That fixed that one!

carlocab commented on Formula/infer.rb:
```diff
@@ -122,13 +89,34 @@ def install
-    shell_output("#{bin}/infer --fail-on-issue -P -- clang -c PassingTest.c")
+    no_issues_output = "\n  No issues found  \n"
+
+    failing_c_output = <<-EOS
+
+FailingTest.c:5: error: Null Dereference
+  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
+  3. int main() {
+  4.   int *s = NULL;
+  5.   *s = 42;
+       ^
+  6.   return 0;
+  7. }
+
+
+Found 1 issue
+          Issue Type(ISSUED_TYPE_ID): #
+  Null Dereference(NULL_DEREFERENCE): 1
+    EOS
```
```suggestion
    failing_c_output = <<~EOS

      FailingTest.c:5: error: Null Dereference
        pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
        3. int main() {
        4.   int *s = NULL;
        5.   *s = 42;
             ^
        6.   return 0;
        7. }


      Found 1 issue
                Issue Type(ISSUED_TYPE_ID): #
        Null Dereference(NULL_DEREFERENCE): 1
    EOS
```
Doesn't this work?

andrewvaughanj replied:
It doesn't appear to:

```
 FailingTest.c:5: error: Null Dereference
-pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
-3. int main() {
-4.   int *s = NULL;
-5.   *s = 42;
-     ^
-6.   return 0;
-7. }
+  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
+  3. int main() {
+  4.   int *s = NULL;
+  5.   *s = 42;
+       ^
+  6.   return 0;
+  7. }


 Found 1 issue
-        Issue Type(ISSUED_TYPE_ID): #
-Null Dereference(NULL_DEREFERENCE): 1
+          Issue Type(ISSUED_TYPE_ID): #
+  Null Dereference(NULL_DEREFERENCE): 1
 "
```

I've tried a few combinations here and none of them seem to work 😕

MikeMcQuaid replied:
Use `<<-EOS` everywhere. That should work unless you need trailing whitespace.

andrewvaughanj replied:
To be clear, the one I tried and pasted here was using `<<~EOS` with *six* leading spaces

carlocab replied:
```suggestion
    failing_c_output = <<~EOS

      FailingTest.c:5: error: Null Dereference
      \ \ pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
      \ \ 3. int main() {
      \ \ 4.   int *s = NULL;
      \ \ 5.   *s = 42;
      \ \      ^
      \ \ 6.   return 0;
      \ \ 7. }


      Found 1 issue
      \ \       Issue Type(ISSUED_TYPE_ID): #
      \ \ Null Dereference(NULL_DEREFERENCE): 1
    EOS
```
Not sure why it's eating all of your leading spaces (it shouldn't be eating spaces after `FailingTest.c`), but you can always escape them.

carlocab replied:
Actually, you probably just need to escape the first space.

andrewvaughanj replied:
```diff
-
-FailingTest.c:5: error: Null Dereference
-  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
-  3. int main() {
-  4.   int *s = NULL;
-  5.   *s = 42;
-       ^
-  6.   return 0;
-  7. }
-
-
-Found 1 issue
-          Issue Type(ISSUED_TYPE_ID): #
-  Null Dereference(NULL_DEREFERENCE): 1
+
+      FailingTest.c:5: error: Null Dereference
+      \  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
+      \  3. int main() {
+      \  4.   int *s = NULL;
+      \  5.   *s = 42;
+      \       ^
+      \  6.   return 0;
+      \  7. }
+
+
+      Found 1 issue
+      \          Issue Type(ISSUED_TYPE_ID): #
+      \  Null Dereference(NULL_DEREFERENCE): 1
```

Gives:

```
Error: infer: failed
An exception occurred within a child process:
  Minitest::Assertion: --- expected
+++ actual
@@ -1,15 +1,15 @@
-"
-      FailingTest.c:5: error: Null Dereference
-        pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
-        3. int main() {
-        4.   int *s = NULL;
-        5.   *s = 42;
-             ^
-        6.   return 0;
-        7. }
-
-
-      Found 1 issue
-                Issue Type(ISSUED_TYPE_ID): #
-        Null Dereference(NULL_DEREFERENCE): 1
+"
+FailingTest.c:5: error: Null Dereference
+  pointer `s` last assigned on line 4 could be null and is dereferenced at line 5, column 3.
+  3. int main() {
+  4.   int *s = NULL;
+  5.   *s = 42;
+       ^
+  6.   return 0;
+  7. }
+
+
+Found 1 issue
+          Issue Type(ISSUED_TYPE_ID): #
+  Null Dereference(NULL_DEREFERENCE): 1
 "
```

Sorry if it seems that I am being hopeless here, I'm really trying to get this "happy", but it seems something isn't playing ball ...

andrewvaughanj replied:
We got there 🥳 

```bash
1 file inspected, no offenses detected
```

andrewvaughanj replied:
Thanks for sticking with me, @carlocab 💪

carlocab commented:
This is probably ready to go as soon as 1.1.0 is registered as the latest release on GitHub. However, if you re-tag 1.1.0 to pull in some last-minute changes, please update the checksum here.

andrewvaughanj commented:
> This is probably ready to go as soon as 1.1.0 is registered as the latest release on GitHub. However, if you re-tag 1.1.0 to pull in some last-minute changes, please update the checksum here, as that will change.

@jvillard ☝️

carlocab commented:
While we're waiting for the new release to land, can you squash all your commits together, with the commit message `infer 1.1.0`?

I'll also flag the ARM runner to attempt a build to see if this will build on ARM. (Or not, if you already know that it doesn't work.)

andrewvaughanj commented:
Squashed; no idea about ARM, sorry!

carlocab commented:
Now we know:
```
Error:  No compiler matching 'ocaml-variants.4.11.1+flambda' found, use 'opam switch list-available' to see what is available, or use '--packages' to select packages explicitly.
```
https://github.com/Homebrew/homebrew-core/pull/73440/checks?check_run_id=2170785058#step:7:3709

Bo98 commented:
It _might_ work if we used brewed OCaml.

carlocab commented:
Or just picked a version that works on Apple Silicon. Pretty sure 4.11.x doesn't.

andrewvaughanj commented:
> However, if you re-tag 1.1.0 to pull in some last-minute changes, please update the checksum here, as that will change.

Okay, there was a new 1.1.0 -- I've updated the checksum, squashed and pushed 🤞

carlocab commented:
Thanks, @andrewvaughanj. Let us know when it's registered as the latest release on GitHub.

Btw, the [Git docs](https://git-scm.com/docs/git-tag#_on_re_tagging) actually have some rather strong words about re-tagging, and how you shouldn't do it. Might be good to keep in mind for the future.

carlocab commented:
Still labelled a pre-release though...

jvillard commented:
My bad, I must have checked that box by mistake. Thanks @carlocab, no longer a pre-release now.

carlocab reviewed (approved):
Thanks, @andrewvaughanj.
